engine/extended_providers/extended_providers.go

```diff
@@ -0,0 +1,164 @@
+package extendedproviders
+
+import (
+	"errors"
+
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// AdBuilder contains fields required for building and signing of a new ad with Extended Providers
+type AdBuilder struct {
+	// providerID contains a peer ID of the main provider (the one from the body of the ad)
+	providerID string
+	// privKey contains a private key of the main provider
+	privKey crypto.PrivKey
+	// addrs contains addresses of the main provider
+	addrs []string
+	// providers contains providers' identities, keys, metadata and addresses that will be used as extended providers in the ad.
+	providers []Info
+	// contextID contains optional context id
+	contextID []byte
+	// contextID contains optional metadata
+	metadata []byte
+	// override contains override flag that is false by default
+	override bool
+	// lastAdID contains optional last ad cid which is cid.Undef by default
+	lastAdID cid.Cid
+}
+
+// Info contains information about extended provider.
+type Info struct {
+	// ID contains peer ID of the extended provider
+	ID string
+	// Metadata contains optional metadata of the extended provider
+	Metadata []byte
+	// Addrs contains a list of extended provider's addresses
+	Addrs []string
+	// Priv contains a provtae key of the extended provider
+	Priv crypto.PrivKey
```

Comment:
```suggestion
	// Priv contains a private key of the extended provider
	Priv crypto.PrivKey
```

are there cases where the publishing process won't have the private keys of the whole family?

Reply:
I thought a lot about it. In our current usecase - allow a provider to scale up by adding new bitswap aliases on other libp2p hosts - the OG provider should have access too all XP keys. 

If going forwards we have such usecase when the main provider and it's XPs are in different security domains, then we can extend the protocol between two different `index-provider`s to be able to send partially signed ads to each other and assemble all required signatures that way.

Reply:
I'd encourage checking with the providers to understand how this works with their current key management strategy - i don't think they have the keys being shared back currently, so if they're going to add a message protocol for comms to support that, we may need to jump to also supporting the partial signing

Reply:
Good point, noted.